File: src/examples/process_explorer/oauth-get-all-records/src/utils/open_api_proxy.js
"use strict";

import axios from "axios";
import * as fflate from "fflate";
import 'dotenv/config';

const CLIENT_ID = process.env.CLIENT_ID;
const CLIENT_SECRET = process.env.CLIENT_SECRET;
const COGNITO_AUTHORIZATION_URL = process.env.COGNITO_AUTHORIZATION_URL;

/**
 * This takes care of making requests to QbDVision's Open API (REST API) using an API key. This tests how clients
 * integrate with QbDVision.
 *
 * @type {exports.OpenAPIProxy}
 */
export class OpenAPIProxy {
    /**
     * @param apiKey {string} The string of digits that QbDVision provides as an API Key.
     * @param baseURL The base URL of the environment the API targets.
     */
    constructor(baseURL) {
        this.baseURL = baseURL;
    }

    async login() {
        try {
            const params = new URLSearchParams();
            params.append('grant_type', 'client_credentials');
            params.append('client_id', CLIENT_ID);
            params.append('client_secret', CLIENT_SECRET);

            const response = await axios.post(
                COGNITO_AUTHORIZATION_URL,
                params,
                {
                    headers: {
                        'Content-Type': 'application/x-www-form-urlencoded'
                    }
                }
            );

            // The access_token in the response is already a JWT
            this.accessToken = response.data.access_token;
            console.log("Received response data: ", response.data);
        } catch (error) {
            console.error('Error:', error);
            throw error;
        }

        // Set up future API calls
        const baseURL = this.baseURL;
        this.axios = axios.create({
            baseURL,
            headers: {
                "Authorization": this.accessToken,
            }
        });
    }

    // Delegate all the API calls to Axios. Read more here: https://axios-http.com/docs/api_intro
    async get(url, config) {
        return await this.decompressIfNeeded(await this.axios.get(url, config));
    }

    async put(url, data, config) {
        return await this.decompressIfNeeded(await this.axios.put(url, data, config));
    }

    async post(url, data, config) {
        return await this.decompressIfNeeded(await this.axios.post(url, data, config));
    }

    async delete(url, config) {
        return await this.decompressIfNeeded(await this.axios.delete(url, config));
    }

    decompressIfNeeded(result) {
        if (result.data?.data) {
            const input = result.data.data;
            // This is required in order to make fflate decompress.
            const fflateCompressedLatinEncodedStr = fflate.strToU8(input, true);

            // Decompress the compressed data
            const decompressedDataStream = fflate.decompressSync(fflateCompressedLatinEncodedStr);

            // Convert the decompressed data stream to string
            const decompressedString = fflate.strFromU8(decompressedDataStream);

            // Parse the string to JSON
            result.data = JSON.parse(decompressedString);
        }

        return result;
    }
}

File: src/examples/process_explorer/oauth-get-all-records-py/src/utils/open_api_proxy.py
import json
import os
import requests
from dotenv import load_dotenv
import gzip
import io

# Load environment variables
load_dotenv()

CLIENT_ID = os.getenv('CLIENT_ID')
CLIENT_SECRET = os.getenv('CLIENT_SECRET')
COGNITO_AUTHORIZATION_URL = os.getenv('COGNITO_AUTHORIZATION_URL')


class OpenAPIProxy:
    """
    This takes care of making requests to QbDVision's Open API (REST API) using OAuth.
    """

    def __init__(self, base_url):
        """
        Initialize the OpenAPIProxy with a base URL.
        
        Args:
            base_url (str): The base URL of the environment the API targets.
        """
        self.base_url = base_url
        self.access_token = None

    def login(self):
        """Authenticate using OAuth 2.0 client credentials flow."""
        try:
            payload = {
                'grant_type': 'client_credentials',
                'client_id': CLIENT_ID,
                'client_secret': CLIENT_SECRET
            }

            headers = {
                'Content-Type': 'application/x-www-form-urlencoded'
            }

            response = requests.post(
                COGNITO_AUTHORIZATION_URL,
                data=payload,
                headers=headers
            )
            response.raise_for_status()

            # The access_token in the response is already a JWT
            data = response.json()
            self.access_token = data['access_token']

        except Exception as error:
            print(f'Error: {error}')
            raise error

    def get_headers(self):
        """Get the headers for API requests with authorization token."""
        return {
            "Authorization": self.access_token
        }

    def get(self, url, params=None):
        """Make a GET request to the API."""
        full_url = f"{self.base_url}{url}"
        response = requests.get(full_url, params=params,
                                headers=self.get_headers())
        response.raise_for_status()
        return self.decompress_if_needed(response)

    def put(self, url, data=None, params=None):
        """Make a PUT request to the API."""
        full_url = f"{self.base_url}{url}"
        response = requests.put(full_url, json=data, params=params,
                                headers=self.get_headers())
        response.raise_for_status()
        return self.decompress_if_needed(response)

    def post(self, url, data=None, params=None):
        """Make a POST request to the API."""
        full_url = f"{self.base_url}{url}"
        response = requests.post(full_url, json=data, params=params,
                                 headers=self.get_headers())
        response.raise_for_status()
        return self.decompress_if_needed(response)

    def delete(self, url, params=None):
        """Make a DELETE request to the API."""
        full_url = f"{self.base_url}{url}"
        response = requests.delete(full_url, params=params,
                                   headers=self.get_headers())
        response.raise_for_status()
        return self.decompress_if_needed(response)

    def decompress_if_needed(self, response):
        """Decompress the response data if it's compressed."""
        result = {"data": response.json()}

        if result["data"] and "data" in result["data"]:
            input_data = result["data"]["data"]

            compressed_data = input_data.encode('latin-1')
            with gzip.GzipFile(fileobj=io.BytesIO(compressed_data)) as f:
                decompressed_data = f.read()

            # Convert to string and parse JSON
            result["data"] = json.loads(decompressed_data.decode('utf-8'))

        return result
